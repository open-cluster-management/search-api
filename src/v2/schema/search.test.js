/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Search Resolver', () => {
  test('Correctly Resolves Search Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchResult: search(input: {
                keywords: [],
                filters: [ { property: "cluster", values: ["cluster1"]}],
                limit: 10000,
                relatedKinds: ["pod"]}){
              items
              related {
                kind
                count
                items
              }
              count
            }
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves Search Keyword Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchResult: search(input: {
                keywords: ["testing"],
                filters: []}){
              items
              related {
                kind
                count
                items
              }
              count
            }
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves Performant Search Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchResult: search(input: {
                keywords: [],
                filters: [ { property: "cluster", values: ["cluster1"]}]}){
              items
              related {
                kind
                items
              }
            }
          }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves SearchSchema Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchSchema
          }
        `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves SearchComplete Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchComplete(property:"kind", query: {keywords:[], filters:[]})
          }
        `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves SearchComplete Query With Filters', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchComplete(property:"storage", query: {keywords:[], filters:[{property:"kind", values:"cluster"}]})
          }
        `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
  test('Correctly Resolves SearchComplete Query With No Params', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
          {
            searchComplete(property:"kind")
          }
        `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});
