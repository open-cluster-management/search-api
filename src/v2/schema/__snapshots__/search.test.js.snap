// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Resolver Correctly Resolves Performant Search Query 1`] = `
Object {
  "data": Object {
    "searchResult": Array [
      Object {
        "items": Array [
          Object {
            "cpu": "30%",
            "endpoint": "",
            "kind": "cluster",
            "memory": "20%",
            "name": "mock-cluster",
            "namespace": "mock-mcm-cluster",
            "nodes": 10,
            "status": "ok",
            "storage": "10%",
          },
          Object {
            "cpu": "30%",
            "endpoint": "",
            "kind": "cluster",
            "memory": "20%",
            "name": "mock-cluster-1",
            "namespace": "mock-mcm-cluster",
            "nodes": 10,
            "status": "ok",
            "storage": "10%",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.0",
            "role": "master",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.1",
            "role": "master",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.2",
            "role": "master",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-1",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-2",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-3",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-4",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
        ],
        "related": Array [
          Object {
            "items": Array [
              Object {
                "count": 1,
                "items": Array [
                  Object {
                    "_uid": "123",
                    "kind": "cluster",
                    "name": "mockCluster",
                    "namespace": "mockCluster-ns",
                  },
                ],
                "kind": "cluster",
              },
            ],
            "kind": "cluster",
          },
        ],
      },
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves Performant Search Query 2`] = `
Object {
  "data": Object {
    "searchResult": Array [
      Object {
        "count": 0,
        "items": Array [],
        "related": Array [
          Object {
            "count": 1,
            "items": Array [
              Object {
                "count": 1,
                "items": Array [
                  Object {
                    "_uid": "123",
                    "kind": "cluster",
                    "name": "mockCluster",
                    "namespace": "mockCluster-ns",
                  },
                ],
                "kind": "cluster",
              },
            ],
            "kind": "cluster",
          },
        ],
      },
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves Search Keyword Query 1`] = `
Object {
  "data": Object {
    "searchResult": Array [
      Object {
        "count": 0,
        "items": Array [],
        "related": Array [
          Object {
            "count": 1,
            "items": Array [
              Object {
                "count": 1,
                "items": Array [
                  Object {
                    "_uid": "123",
                    "kind": "cluster",
                    "name": "mockCluster",
                    "namespace": "mockCluster-ns",
                  },
                ],
                "kind": "cluster",
              },
            ],
            "kind": "cluster",
          },
        ],
      },
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves Search Query 1`] = `
Object {
  "data": Object {
    "searchResult": Array [
      Object {
        "count": 10,
        "items": Array [
          Object {
            "cpu": "30%",
            "endpoint": "",
            "kind": "cluster",
            "memory": "20%",
            "name": "mock-cluster",
            "namespace": "mock-mcm-cluster",
            "nodes": 10,
            "status": "ok",
            "storage": "10%",
          },
          Object {
            "cpu": "30%",
            "endpoint": "",
            "kind": "cluster",
            "memory": "20%",
            "name": "mock-cluster-1",
            "namespace": "mock-mcm-cluster",
            "nodes": 10,
            "status": "ok",
            "storage": "10%",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.0",
            "role": "master",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.1",
            "role": "master",
          },
          Object {
            "cpus": 10,
            "kind": "node",
            "name": "mock-1.1.1.2",
            "role": "master",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-1",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-2",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-3",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
          Object {
            "creationTimestamp": "2018-09-27 16:52:20 -0400 EDT",
            "hostIP": "179.160.35.59",
            "kind": "pod",
            "name": "mock-cluster-4",
            "namespace": "kube-system",
            "podIP": "179.160.35.59",
            "restarts": 20,
            "status": "Running",
          },
        ],
        "related": Array [
          Object {
            "count": 1,
            "items": Array [
              Object {
                "count": 1,
                "items": Array [
                  Object {
                    "_uid": "123",
                    "kind": "cluster",
                    "name": "mockCluster",
                    "namespace": "mockCluster-ns",
                  },
                ],
                "kind": "cluster",
              },
            ],
            "kind": "cluster",
          },
        ],
      },
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves SearchComplete Query 1`] = `
Object {
  "data": Object {
    "searchComplete": Array [
      "cluster",
      "application",
      "deployable",
      "compliance",
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves SearchComplete Query 2`] = `
Object {
  "data": Object {
    "searchSchema": Object {
      "allProperties": Array [
        "kind",
        "name",
        "namespace",
        "cpu",
        "created",
      ],
    },
  },
}
`;

exports[`Search Resolver Correctly Resolves SearchComplete Query With Filters 1`] = `
Object {
  "data": Object {
    "searchComplete": Array [
      "kind",
      "name",
      "namespace",
      "cpu",
      "created",
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves SearchComplete Query With No Params 1`] = `
Object {
  "data": Object {
    "searchComplete": Array [
      "cluster",
      "application",
      "deployable",
      "compliance",
    ],
  },
}
`;

exports[`Search Resolver Correctly Resolves SearchSchema Query 1`] = `
Object {
  "data": Object {
    "searchSchema": Object {
      "allProperties": Array [
        "kind",
        "name",
        "namespace",
        "cpu",
        "created",
      ],
    },
  },
}
`;
